pkg/await/apps_statefulset_test.go

```diff
@@ -0,0 +1,1252 @@
+package await
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/watch"
+)
+
+func Test_Apps_StatefulSet(t *testing.T) {
+	const (
+		inputNamespace = "default"
+		inputName      = "foo"
+		targetService  = "foo-service"
+	)
+	tests := []struct {
+		description   string
+		do            func(statefulsets, pods chan watch.Event, timeout chan time.Time)
+		expectedError error
+	}{
+		{
+			description: "[Revision 1] Should succeed after creating StatefulSet",
+			do: func(statefulsets, pods chan watch.Event, timeout chan time.Time) {
+				// API server successfully creates and initializes StatefulSet object.
+				statefulsets <- watchAddedEvent(
+					statefulsetAdded(inputNamespace, inputName, targetService))
+				statefulsets <- watchAddedEvent(
+					statefulsetCreating(inputNamespace, inputName, targetService))
+				statefulsets <- watchAddedEvent(
+					statefulsetProgressing(inputNamespace, inputName, targetService))
+				statefulsets <- watchAddedEvent(
+					statefulsetReady(inputNamespace, inputName, targetService))
+
+				// Timeout. Success.
+				timeout <- time.Now()
+			},
+		},
+		{
+			description: "[Revision 2] Should succeed after updating StatefulSet",
+			do: func(statefulsets, pods chan watch.Event, timeout chan time.Time) {
+				// API server successfully updates StatefulSet object.
+				statefulsets <- watchAddedEvent(
+					statefulsetUpdate(inputNamespace, inputName, targetService))
+				statefulsets <- watchAddedEvent(
+					statefulsetUpdating(inputNamespace, inputName, targetService))
+				statefulsets <- watchAddedEvent(
+					statefulsetUpdatingWithActiveReplica(inputNamespace, inputName, targetService))
+				statefulsets <- watchAddedEvent(
+					statefulsetUpdateSuccess(inputNamespace, inputName, targetService))
+
+				// Timeout. Success.
+				timeout <- time.Now()
+			},
+		},
+		{
+			description: "[Revision 1] Should fail if timeout occurs before successful creation",
+			do: func(statefulsets, pods chan watch.Event, timeout chan time.Time) {
+				// API server successfully creates StatefulSet object.
+				statefulsets <- watchAddedEvent(
+					statefulsetAdded(inputNamespace, inputName, targetService))
+				statefulsets <- watchAddedEvent(
+					statefulsetCreating(inputNamespace, inputName, targetService))
+
+				// Timeout. Failure.
+				timeout <- time.Now()
+			},
+			expectedError: &timeoutError{
+				object: statefulsetCreating(inputNamespace, inputName, targetService),
+				subErrors: []string{
+					"Failed to observe the expected number of ready replicas",
```

Comment:
1. Is my recollection correct, that in cases where we've timed out or cancelled, we actually say that we timed out or cancelled? I believe we do and this is actually communicated by the CLI's rendering of the timerout/cancellation error types, but I just want to make 100% sure because I think that's important to communicate.
1. For the scenarios where we fail with only this message -- how often do they occur? IMHO we should always try very, _very_ hard to provide more information if we can, and if this happens relatively frequently I think that's a sign we might want to file an issue to follow up, at least.

Reply:
1. Yes, that's right.
2. I think in practice, you'd also get Pod failure messages. Here's a real example:
```
Diagnostics:
  kubernetes:apps:StatefulSet (statefulset):
    error: Plan apply failed: 4 errors occurred:

    * Timeout occurred for 'foo'
    * Failed to observe the expected number of ready replicas
    * .status.currentRevision does not match .status.updateRevision
    * 1 Pods failed to run because: [ImagePullBackOff] Back-off pulling image "nginx:busted"
```